File: services/blog-service/src/main/java/net/blueshell/blogservice/mapper/SocialMapper.java
package net.blueshell.blogservice.mapper;

import net.blueshell.blogservice.model.Blog;
import net.blueshell.dto.InternalBlogDTO;
import net.blueshell.dto.SocialDTO;
import net.blueshell.enums.PlatformType;
import org.mapstruct.AfterMapping;
import org.mapstruct.Mapper;
import org.mapstruct.Mapping;
import org.mapstruct.MappingTarget;
import org.springframework.beans.factory.annotation.Value;

@Mapper(componentModel = "spring")
public abstract class SocialMapper {

    @Value("${frontend.url}")
    private String frontendUrl;

    public abstract SocialDTO toSocialDTO(Blog blog);

    @AfterMapping
    protected void afterToSocialDTO(Blog blog, @MappingTarget SocialDTO dto) {
        dto.setUrl(frontendUrl + "/blogs" + blog.getId());
        dto.setTitle(blog.getTitle());
        dto.setText(dto.getText());
        PlatformType[] platforms = {PlatformType.FACEBOOK, PlatformType.TWITTER, PlatformType.INSTAGRAM, PlatformType.LINKEDIN};
        dto.setPlatforms(platforms);
    }
}

File: api/src/main/kotlin/net/blueshell/api/blog/api/mapper/SocialMapper.kt
package net.blueshell.api.blog.api.mapper

import net.blueshell.api.shared.enums.PlatformType
import net.blueshell.api.blog.api.dto.SocialDTO
import net.blueshell.api.blog.domain.model.Blog
import org.mapstruct.AfterMapping
import org.mapstruct.BeanMapping
import org.mapstruct.Mapper
import org.mapstruct.MappingTarget
import org.springframework.beans.factory.annotation.Value

@Mapper(componentModel = "spring")
abstract class SocialMapper {
    @Value($$"${frontend.url}")
    private lateinit var frontendUrl: String

    @BeanMapping(ignoreByDefault = true)
    abstract fun toSocialDTO(blog: Blog): SocialDTO

    @AfterMapping
    protected fun afterToSocialDTO(blog: Blog, @MappingTarget dto: SocialDTO) {
        dto.url = frontendUrl + "/blogs" + blog.id
        dto.title = blog.title
        dto.text = dto.text
        val platforms = arrayOf(
            PlatformType.FACEBOOK,
            PlatformType.TWITTER,
            PlatformType.INSTAGRAM,
            PlatformType.LINKEDIN
        )
        dto.platforms = platforms
    }
}
